File: code/ds/c版本/BinaryInsertSort.cpp
// 折半查找
void BinaryInsertSort(ElemType Arr[],int n){
  int i,j,lowIndex,highIndex,midIndex;

  for(i=2;j<=n;i++){
    // 将待排序的元素暂存在Arr[0]上
    Arr[0]=Arr[i];

    lowIndex=1;         // 左侧子表 折半查找起始位置
    highIndex=i-1;     // 左侧子表 折半查找结束位置
    while(lowIndex<=highIndex){

      // 左侧有序子表的中间位置角标
      midIndex=(lowIndex+heightIndex)/2;

      if(Arr[midIndex].key>Arr[0].key){
        // 小于中间元素，插入位置在子表左侧
        highIndex=mid-1
      }else{
        // 大于或者等于中间元素，插入位置在子表右侧
        lowIndex=midIndex+1;
      }
    }

    // 跳出循环需要（lowIndex>heightIndex），
    //  说明待插入位置的角标在heightIndex之后，为 heightIndex+1,此时需要将（heightIndex，i）之间的所有元素后移

    for(j=i-1;j>highIndex;--j){
      Arr[j+1]=Arr[j]
    }

    // 后移完成后，将元素Arr[0]赋值到位置（highIndex+1）上
    Arr[highIndex+1]=Arr[0]
  }
}

File: code/ds/ts版本/BinaryInsertSort.ts
/**
 * 折半插入排序
 */
function binaryInsertSort(arr: number[], len: number): number[] {
  // 数组长度校验【非必须】
  if (arr.length !== len) {
    len = arr.length
  }

  // 遍历
  for (let i = 1; i < len; i++) {
    const temp = arr[i]
    let lowIndex = 0
    let highIndex = i - 1

    while (lowIndex <= highIndex) {
      // 注意：取整,javascript这里取整，会出现空指针
      const mid = Math.ceil((lowIndex + highIndex) / 2)

      if (arr[mid] <= temp) {
        // 右侧
        lowIndex = mid + 1
      }
      else {
        // 左侧
        highIndex = mid - 1
      }
    }
    // 元素后移
    for (let j = i - 1; j > highIndex; --j) {
      arr[j + 1] = arr[j]
    }

    arr[highIndex + 1] = temp
  }

  // 返回经过折半插入排序处理的有序数组
  return arr
}

// 测试用例
const dealArr = [5, 2, 7, 3, 18, 8, 12, 1]
console.log('插入排序前：', dealArr)

const sortResult = binaryInsertSort(dealArr, 7)
console.log('插入排序后：', sortResult)
